File: audit-service/controllers/auditController.js
const db = require("../config/db");
require("dotenv").config();
const axios = require("axios");
const transporter = require("../services/mailer");

// Función para obtener el companyId de un usuario
async function getCompanyIdForUser(userId) {
  try {
    const userResponse = await axios.get(
      `${process.env.USER_SERVICE_URL}/${userId}/company`
    );
    const companyId = userResponse.data.company_id;
    return companyId;
  } catch (error) {
    console.error("Error obteniendo el companyId", error);
    return null;
  }
}

// Función para enviar correo de notificación
async function sendEmailNotification(companyId, featureName, values) {
  // Obtener los usuarios suscritos a la empresa
  const users = await getUsersByCompanyId(companyId); // Función para obtener los usuarios suscritos
  const subscribedUsers = users.filter((user) => user.is_subscribed); // Filtramos los suscritos

  // Enviar correos a los usuarios suscritos
  for (const user of subscribedUsers) {
    const mailOptions = {
      from: process.env.EMAIL_USER,
      to: user.email,
      subject: `Notificación de actualización en feature: ${featureName}`,
      text: `Hola ${user.first_name},\n\nLa feature "${featureName}" ha sido actualizada.\n "${values}" \n Te mantenemos informado.\n\nSaludos,\nTu equipo.`,
    };

    transporter.sendMail(mailOptions, (error, info) => {
      if (error) {
        console.error("Error al enviar el correo:", error);
      } else {
        console.log("Correo enviado:", info.response);
      }
    });
  }
}

// Función para obtener los usuarios suscritos a una empresa
async function getUsersByCompanyId(companyId) {
  try {
    // Hacer la solicitud GET al microservicio de usuarios
    const response = await axios.get(
      `${process.env.USER_SERVICE_URL}/${companyId}/users`
    );
    console.log("Usuarios de la empresa:", response.data);
    // Si la respuesta es exitosa, retornar los datos de los usuarios
    return response.data;
  } catch (error) {
    console.error("Error al obtener usuarios de la empresa:", error);

    // Si hay error en la llamada, lanzar un error para manejarlo en el controlador
    throw new Error("No se pudieron obtener los usuarios de la empresa.");
  }
}

// Función para registrar una acción de auditoría con promesas (usando async/await)
async function logAuditAction(action, entity, entityId, details, userId) {
  const timestamp = new Date().toISOString(); // Fecha y hora actual

  const auditRecord = {
    action,
    entity,
    entity_id: entityId,
    details: JSON.stringify(details),
    user_id: userId,
    timestamp,
  };

  const query =
    "INSERT INTO audit_log (action, entity, entity_id, details, user_id, timestamp) VALUES (?, ?, ?, ?, ?, ?)";
  const values = [
    auditRecord.action,
    auditRecord.entity,
    auditRecord.entity_id,
    auditRecord.details,
    auditRecord.user_id,
    auditRecord.timestamp,
  ];

  try {
    const [result] = await db.promise().execute(query, values);

    // Si la entidad es 'feature', enviamos la notificación
    if (auditRecord.entity === "feature") {
      // Obtener el companyId del usuario
      const companyId = await getCompanyIdForUser(auditRecord.user_id);
      try {
        // Enviar la notificación a todos los usuarios suscritos
        await sendEmailNotification(companyId, auditRecord.entity_id, values);
      } catch (error) {
        console.error("Error enviando notificación por correo:", error);
      }
    }
  } catch (err) {
    console.error("Error al registrar la auditoría:", err);
  }
}

// Función para obtener los registros de auditoría con filtros usando async/await
async function getAuditLogs(filters) {
  let queryBase = "SELECT * FROM audit_log WHERE 1 = 1"; // Iniciar consulta con filtro de todas las entradas
  const queryValues = [];

  // Agregar filtros
  if (filters.startDate && filters.endDate) {
    queryBase += " AND timestamp BETWEEN ? AND ?";
    queryValues.push(filters.startDate, filters.endDate);
  }

  if (filters.action) {
    queryBase += " AND action = ?";
    queryValues.push(filters.action);
  }

  if (filters.userId) {
    queryBase += " AND user_id = ?";
    queryValues.push(filters.userId);
  }

  try {
    const [rows] = await db.promise().execute(queryBase, queryValues);
    return rows; // Retorna los registros obtenidos
  } catch (err) {
    console.error("Error al obtener los registros de auditoría:", err);
    throw new Error("Error al obtener los registros de auditoría");
  }
}

module.exports = { logAuditAction, getAuditLogs };

File: audit-service/controllers/auditController.py
import os
import json
import logging
import boto3
from flask import jsonify
from datetime import datetime
import pymysql
import requests
from dotenv import load_dotenv
load_dotenv()

# Configuración del cliente SQS
sqs = boto3.client(
    "sqs",
    region_name=os.getenv("AWS_REGION"),
    aws_access_key_id=os.getenv("AWS_ACCESS_KEY_ID"),
    aws_secret_access_key=os.getenv("AWS_SECRET_ACCESS_KEY"),
    aws_session_token = os.getenv("AWS_SESSION_TOKEN")
)

# Configuración de la conexión a la base de datos
db_connection = pymysql.connect(
    host=os.getenv("RDS_HOSTNAME"),
    user=os.getenv("RDS_USERNAME"),
    password=os.getenv("RDS_PASSWORD"),
    database=os.getenv("RDS_DATABASE"),
    cursorclass=pymysql.cursors.DictCursor,
)

# Función para obtener el companyId de un usuario
def get_company_id_for_user(user_id):
    try:
        response = requests.get(f"{os.getenv('USER_SERVICE_URL')}/{user_id}/company")
        response.raise_for_status()
        company_id = response.json().get("company_id")
        return company_id
    except requests.RequestException as e:
        logging.error(f"Error obteniendo el companyId: {e}")
        return None

# Función para obtener los usuarios por company_id
def get_users_by_company_id(company_id):
    """Obtiene los usuarios de la empresa desde el servicio de usuarios."""
    try:
        response = requests.get(f"{os.getenv('USER_SERVICE_URL')}/{company_id}/users")
        response.raise_for_status()
        return response.json()
    except requests.RequestException as e:
        logging.error(f"Error al obtener usuarios de la empresa {company_id}: {e}")
        return []  # Devuelve una lista vacía para evitar errores en el flujo

# Función para registrar una acción de auditoría
def log_audit_action(action, entity, entity_id, details, user_id):
    timestamp = datetime.utcnow().strftime("%Y-%m-%d %H:%M:%S")
    query = """
        INSERT INTO audit_log (action, entity, entity_id, details, user_id, timestamp)
        VALUES (%s, %s, %s, %s, %s, %s)
    """
    values = (action, entity, entity_id, json.dumps(details), user_id, timestamp)

    try:
        with db_connection.cursor() as cursor:
            cursor.execute(query, values)
            db_connection.commit()
    except Exception as e:
        logging.error(f"Error al registrar la auditoría: {e}")
        db_connection.rollback()  # Hacer rollback en caso de error
        raise  # Re-lanzar la excepción para gestionarla arriba

    # Enviar mensaje a la cola de SQS si la entidad es "feature"
    if entity == "feature":
        company_id = get_company_id_for_user(user_id)
        if company_id:
            # Obtener los usuarios de la empresa
            users = get_users_by_company_id(company_id)
            if users:
                # Crear el mensaje para enviar a la cola SQS
                message_body = {
                    "company_id": company_id,
                    "feature_name": entity_id,
                    "values": details,
                    "users": users,  # Incluir usuarios en el mensaje
                }
                try:
                    sqs.send_message(
                        QueueUrl=os.getenv("SQS_QUEUE_URL"),
                        MessageBody=json.dumps(message_body),
                    )
                    print("Mensaje enviado a la cola de SQS.")
                    logging.info("Mensaje enviado a la cola de SQS.")
                except Exception as sqs_error:
                    logging.error(f"Error al enviar el mensaje a la cola de SQS: {sqs_error}")

# Función para obtener registros de auditoría con filtros
def get_audit_logs(filters):
    query_base = "SELECT * FROM audit_log WHERE 1=1"
    query_values = []

    if "startDate" in filters and "endDate" in filters:
        query_base += " AND timestamp BETWEEN %s AND %s"
        query_values.extend([filters["startDate"], filters["endDate"]])

    if "action" in filters:
        query_base += " AND action = %s"
        query_values.append(filters["action"])

    if "userId" in filters:
        query_base += " AND user_id = %s"
        query_values.append(filters["userId"])

    try:
        with db_connection.cursor() as cursor:
            cursor.execute(query_base, tuple(query_values))
            return cursor.fetchall()  # Devuelve los registros encontrados
    except Exception as e:
        logging.error(f"Error al obtener los registros de auditoría: {e}")
        raise ValueError("Error al obtener los registros de auditoría")

